// @botonic/core/lib/esm/models/events/message/message-event.js
export var MessageEventTypes;
(function (MessageEventTypes) {
    MessageEventTypes["AUDIO"] = "audio";
    MessageEventTypes["CAROUSEL"] = "carousel";
    MessageEventTypes["CUSTOM"] = "custom";
    MessageEventTypes["DOCUMENT"] = "document";
    MessageEventTypes["IMAGE"] = "image";
    MessageEventTypes["LOCATION"] = "location";
    MessageEventTypes["TEXT"] = "text";
    MessageEventTypes["POSTBACK"] = "postback";
    MessageEventTypes["VIDEO"] = "video";
    /**
     * TODO: contact, buttonmessage?, webchatsettings, whatsapp template
     */
})(MessageEventTypes || (MessageEventTypes = {}));
export var MessageEventAck;
(function (MessageEventAck) {
    MessageEventAck["DRAFT"] = "draft";
    MessageEventAck["READ"] = "read";
    MessageEventAck["RECEIVED"] = "received";
    MessageEventAck["SENT"] = "sent";
})(MessageEventAck || (MessageEventAck = {}));
export var MessageEventFrom;
(function (MessageEventFrom) {
    MessageEventFrom["AGENT"] = "agent";
    MessageEventFrom["BOT"] = "bot";
    MessageEventFrom["USER"] = "user";
})(MessageEventFrom || (MessageEventFrom = {}));
