// @botonic/react/lib/webchat/components/attachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Attachment = void 0;

var _react = _interopRequireWildcard(require("react"));

var _attachmentIcon = _interopRequireDefault(require("../../assets/attachment-icon.svg"));

var _constants = require("../../constants");

var _contexts = require("../../contexts");

var _conditionalAnimation = require("../components/conditional-animation");

var _common = require("./common");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Attachment = function Attachment(_ref) {
  var onChange = _ref.onChange,
      accept = _ref.accept,
      enableAttachments = _ref.enableAttachments;

  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      getThemeProperty = _useContext.getThemeProperty;

  var CustomAttachments = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.customAttachments, undefined);

  var isAttachmentsEnabled = function isAttachmentsEnabled() {
    var _getThemeProperty;

    var hasCustomAttachments = !!CustomAttachments;
    return (_getThemeProperty = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.enableAttachments, enableAttachments)) !== null && _getThemeProperty !== void 0 ? _getThemeProperty : hasCustomAttachments;
  };

  var attachmentsEnabled = isAttachmentsEnabled();
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, attachmentsEnabled ? /*#__PURE__*/_react["default"].createElement(_conditionalAnimation.ConditionalAnimation, null, /*#__PURE__*/_react["default"].createElement("div", {
    role: _constants.ROLES.ATTACHMENT_ICON
  }, /*#__PURE__*/_react["default"].createElement("label", {
    htmlFor: "attachment",
    style: {
      marginTop: 4
    }
  }, CustomAttachments ? /*#__PURE__*/_react["default"].createElement(CustomAttachments, null) : /*#__PURE__*/_react["default"].createElement(_common.Icon, {
    src: _attachmentIcon["default"]
  })), /*#__PURE__*/_react["default"].createElement("input", {
    type: "file",
    name: "file",
    id: "attachment",
    style: {
      display: 'none'
    },
    onChange: onChange,
    accept: accept
  }))) : null);
};

exports.Attachment = Attachment;
