// @botonic/react/lib/webchat/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UPDATE_WEBVIEW = exports.UPDATE_TYPING = exports.UPDATE_THEME = exports.UPDATE_SESSION = exports.UPDATE_REPLIES = exports.UPDATE_MESSAGE = exports.UPDATE_LATEST_INPUT = exports.UPDATE_LAST_ROUTE_PATH = exports.UPDATE_LAST_MESSAGE_DATE = exports.UPDATE_JWT = exports.UPDATE_HANDOFF = exports.UPDATE_DEV_SETTINGS = exports.TOGGLE_WEBCHAT = exports.TOGGLE_PERSISTENT_MENU = exports.TOGGLE_EMOJI_PICKER = exports.TOGGLE_COVER_COMPONENT = exports.SET_ONLINE = exports.SET_ERROR = exports.SET_CURRENT_ATTACHMENT = exports.DO_RENDER_CUSTOM_COMPONENT = exports.CLEAR_MESSAGES = exports.ADD_MESSAGE_COMPONENT = exports.ADD_MESSAGE = void 0;
var ADD_MESSAGE = 'addMessage';
exports.ADD_MESSAGE = ADD_MESSAGE;
var ADD_MESSAGE_COMPONENT = 'addMessageComponent';
exports.ADD_MESSAGE_COMPONENT = ADD_MESSAGE_COMPONENT;
var UPDATE_MESSAGE = 'updateMessage';
exports.UPDATE_MESSAGE = UPDATE_MESSAGE;
var UPDATE_REPLIES = 'updateReplies';
exports.UPDATE_REPLIES = UPDATE_REPLIES;
var UPDATE_LATEST_INPUT = 'updateLatestInput';
exports.UPDATE_LATEST_INPUT = UPDATE_LATEST_INPUT;
var UPDATE_TYPING = 'updateTyping';
exports.UPDATE_TYPING = UPDATE_TYPING;
var UPDATE_WEBVIEW = 'updateWebview';
exports.UPDATE_WEBVIEW = UPDATE_WEBVIEW;
var UPDATE_SESSION = 'updateSession';
exports.UPDATE_SESSION = UPDATE_SESSION;
var UPDATE_LAST_ROUTE_PATH = 'updateLastRoutePath';
exports.UPDATE_LAST_ROUTE_PATH = UPDATE_LAST_ROUTE_PATH;
var UPDATE_HANDOFF = 'updateHandoff';
exports.UPDATE_HANDOFF = UPDATE_HANDOFF;
var UPDATE_THEME = 'updateTheme';
exports.UPDATE_THEME = UPDATE_THEME;
var UPDATE_DEV_SETTINGS = 'updateDevSettings';
exports.UPDATE_DEV_SETTINGS = UPDATE_DEV_SETTINGS;
var TOGGLE_WEBCHAT = 'toggleWebchat';
exports.TOGGLE_WEBCHAT = TOGGLE_WEBCHAT;
var TOGGLE_EMOJI_PICKER = 'toggleEmojiPicker';
exports.TOGGLE_EMOJI_PICKER = TOGGLE_EMOJI_PICKER;
var TOGGLE_PERSISTENT_MENU = 'togglePersistentMenu';
exports.TOGGLE_PERSISTENT_MENU = TOGGLE_PERSISTENT_MENU;
var TOGGLE_COVER_COMPONENT = 'toggleCoverComponent';
exports.TOGGLE_COVER_COMPONENT = TOGGLE_COVER_COMPONENT;
var DO_RENDER_CUSTOM_COMPONENT = 'doRenderCustomComponent';
exports.DO_RENDER_CUSTOM_COMPONENT = DO_RENDER_CUSTOM_COMPONENT;
var SET_ERROR = 'setError';
exports.SET_ERROR = SET_ERROR;
var CLEAR_MESSAGES = 'clearMessages';
exports.CLEAR_MESSAGES = CLEAR_MESSAGES;
var UPDATE_LAST_MESSAGE_DATE = 'updateLastMessageDate';
exports.UPDATE_LAST_MESSAGE_DATE = UPDATE_LAST_MESSAGE_DATE;
var SET_CURRENT_ATTACHMENT = 'setCurrentAttachment';
exports.SET_CURRENT_ATTACHMENT = SET_CURRENT_ATTACHMENT;
var SET_ONLINE = 'setOnline';
exports.SET_ONLINE = SET_ONLINE;
var UPDATE_JWT = 'updateJwt';
exports.UPDATE_JWT = UPDATE_JWT;
