// @botonic/react/lib/webchat/header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebchatHeader = exports.StyledWebchatHeader = exports.DefaultHeader = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _framerMotion = require("framer-motion");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _contexts = require("../contexts");

var _environment = require("../util/environment");

var _react2 = require("../util/react");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Header = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  background: linear-gradient(\n    90deg,\n    ", " 0%,\n    ", " 100%\n  );\n  height: 55px;\n  border-radius: ", ";\n  z-index: 2;\n"])), _constants.COLORS.BLEACHED_CEDAR_PURPLE, function (props) {
  return props.color;
}, _constants.WEBCHAT.DEFAULTS.BORDER_RADIUS_TOP_CORNERS);

var ImageContainer = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 10px;\n  align-items: center;\n"])));

var Image = _styledComponents["default"].img(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  width: 32px;\n  border-radius: 50%;\n"])));

var TextContainer = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n  flex: 1 1 auto;\n"])));

var Title = _styledComponents["default"].div(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  font-family: inherit;\n  font-size: 15px;\n  font-weight: bold;\n  color: ", ";\n"])), _constants.COLORS.SOLID_WHITE);

var Subtitle = _styledComponents["default"].div(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  font-family: inherit;\n  font-size: 11px;\n  color: ", ";\n"])), _constants.COLORS.SOLID_WHITE);

var CloseHeader = _styledComponents["default"].div(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 0px 16px;\n  cursor: pointer;\n  color: ", ";\n  font-family: inherit;\n  font-size: 36px;\n"])), _constants.COLORS.SOLID_WHITE);

var DefaultHeader = function DefaultHeader(props) {
  var getThemeProperty = props.getThemeProperty;
  var animationsEnabled = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.enableAnimations, true);
  var headerImage = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.headerImage, getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.brandImage, _constants.WEBCHAT.DEFAULTS.LOGO));
  var headerTitle = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.headerTitle, _constants.WEBCHAT.DEFAULTS.TITLE);
  var headerSubtitle = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.headerSubtitle, '');
  return /*#__PURE__*/_react["default"].createElement(Header, {
    role: _constants.ROLES.HEADER,
    color: props.color,
    style: _objectSpread({}, getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.headerStyle))
  }, headerImage && /*#__PURE__*/_react["default"].createElement(ImageContainer, null, /*#__PURE__*/_react["default"].createElement(Image, {
    src: (0, _environment.resolveImage)(headerImage)
  })), /*#__PURE__*/_react["default"].createElement(TextContainer, {
    ml: headerImage ? '0px' : '16px'
  }, /*#__PURE__*/_react["default"].createElement(Title, {
    mb: headerSubtitle ? '6px' : '0px'
  }, headerTitle), /*#__PURE__*/_react["default"].createElement(Subtitle, null, headerSubtitle)), /*#__PURE__*/_react["default"].createElement(_react2.ConditionalWrapper, {
    condition: animationsEnabled,
    wrapper: function wrapper(children) {
      return /*#__PURE__*/_react["default"].createElement(_framerMotion.motion.div, {
        whileHover: {
          scale: 1.2
        }
      }, children);
    }
  }, /*#__PURE__*/_react["default"].createElement(CloseHeader, {
    onClick: props.onCloseClick
  }, "\u2A2F")));
};

exports.DefaultHeader = DefaultHeader;

var WebchatHeader = function WebchatHeader(props) {
  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      webchatState = _useContext.webchatState,
      getThemeProperty = _useContext.getThemeProperty;

  var handleCloseWebchat = function handleCloseWebchat(event) {
    props.onCloseClick(event.target.value);
  };

  var CustomHeader = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.customHeader);

  if (CustomHeader) {
    return /*#__PURE__*/_react["default"].createElement(CustomHeader, {
      onCloseClick: handleCloseWebchat
    });
  }

  return /*#__PURE__*/_react["default"].createElement(DefaultHeader, {
    webchatState: webchatState,
    getThemeProperty: getThemeProperty,
    color: getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.brandColor, _constants.COLORS.BOTONIC_BLUE),
    onCloseClick: handleCloseWebchat
  });
};

exports.WebchatHeader = WebchatHeader;
var StyledWebchatHeader = (0, _styledComponents["default"])(WebchatHeader)(_templateObject8 || (_templateObject8 = (0, _taggedTemplateLiteral2["default"])(["\n  border-radius: 8px 8px 0px 0px;\n  box-shadow: ", " 0px 2px 5px;\n  height: 36px;\n  flex: none;\n"])), _constants.COLORS.PIGEON_POST_BLUE_ALPHA_0_5);
exports.StyledWebchatHeader = StyledWebchatHeader;
