// @botonic/react/lib/webchat/webview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebviewHeader = exports.WebviewContainer = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _reactFrameComponent = _interopRequireDefault(require("react-frame-component"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _contexts = require("../contexts");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var StyledWebview = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  position: absolute;\n  display: flex;\n  flex-direction: column;\n  bottom: 0;\n  width: 100%;\n  height: 100%;\n  background-color: ", ";\n  z-index: 2;\n  border-radius: ", ";\n"])), _constants.COLORS.SOLID_WHITE, _constants.WEBCHAT.DEFAULTS.BORDER_RADIUS_TOP_CORNERS);

var StyledWebviewHeader = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  text-align: right;\n  background-color: ", ";\n  border-top: 1px solid ", ";\n  border-bottom: 1px solid ", ";\n  border-radius: ", ";\n"])), _constants.COLORS.WILD_SAND_WHITE, _constants.COLORS.SOLID_BLACK_ALPHA_0_2, _constants.COLORS.SOLID_BLACK_ALPHA_0_2, _constants.WEBCHAT.DEFAULTS.BORDER_RADIUS_TOP_CORNERS);

var StyledCloseHeader = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  display: inline-block;\n  padding: 8px 12px;\n  cursor: pointer;\n"])));

var StyledWebviewContent = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  flex: 1;\n  overflow: auto;\n"])));

var StyledFrame = _styledComponents["default"].iframe(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  border-style: none;\n  width: 100%;\n  height: 100%;\n"])));

var WebviewMode = function WebviewMode(props) {
  /*
    Default mode is with divs.
    Setting the prop "asframe" will render the webview inside an iframe.
    Pros and Cons of this "asframe" mode are:
    Pros: OAuth2 flows can be tested locally with an iframe.
    Cons: We won't be able to visualize correctly css styles in botonic serve (although styles will work in production).
   */
  var style = {
    borderStyle: 'none',
    width: '100%',
    height: '100%'
  };

  if (props.asframe) {
    return /*#__PURE__*/_react["default"].createElement(_reactFrameComponent["default"], {
      style: style
    }, props.children);
  }

  return /*#__PURE__*/_react["default"].createElement("div", {
    style: style
  }, props.children);
};

var WebviewHeader = function WebviewHeader() {
  var _useContext = (0, _react.useContext)(_contexts.RequestContext),
      closeWebview = _useContext.closeWebview;

  var _useContext2 = (0, _react.useContext)(_contexts.WebchatContext),
      getThemeProperty = _useContext2.getThemeProperty;

  return /*#__PURE__*/_react["default"].createElement(StyledWebviewHeader, {
    role: _constants.ROLES.WEBVIEW_HEADER,
    style: _objectSpread({}, getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.webviewHeaderStyle))
  }, /*#__PURE__*/_react["default"].createElement(StyledCloseHeader, {
    onClick: closeWebview
  }, "\u2715"));
};

exports.WebviewHeader = WebviewHeader;

var WebviewContainer = function WebviewContainer(props) {
  var _useContext3 = (0, _react.useContext)(_contexts.WebchatContext),
      webchatState = _useContext3.webchatState;

  var _useContext4 = (0, _react.useContext)(_contexts.RequestContext),
      closeWebview = _useContext4.closeWebview;

  var Webview = webchatState.webview;

  var close = function close(e) {
    return e.data == 'botonicCloseWebview' && closeWebview();
  };

  (0, _react.useEffect)(function () {
    if (window.addEventListener) {
      window.addEventListener('message', close, false);
    } else if (window.attachEvent) {
      // ie8
      window.attachEvent('onmessage', close);
    }
  }, []);
  return /*#__PURE__*/_react["default"].createElement(StyledWebview, {
    role: _constants.ROLES.WEBVIEW,
    style: _objectSpread({}, props.style)
  }, /*#__PURE__*/_react["default"].createElement(WebviewHeader, {
    style: {
      flex: 'none'
    }
  }), /*#__PURE__*/_react["default"].createElement(StyledWebviewContent, null, typeof Webview === 'string' ? /*#__PURE__*/_react["default"].createElement(StyledFrame, {
    src: Webview
  }) : /*#__PURE__*/_react["default"].createElement(WebviewMode, null, /*#__PURE__*/_react["default"].createElement(Webview, null))));
};

exports.WebviewContainer = WebviewContainer;
