// @botonic/nlp/lib/embeddings/database/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DB_COLUMN_NAME = exports.DB_TABLE_NAME = exports.EMBEDDINGS_URL = exports.GLOBAL_EMBEDDINGS_PATH = exports.SUPPORTED_EMBEDDINGS = void 0;
var os_1 = require("os");
var path_1 = require("path");
var constants_1 = require("../../constants");
exports.SUPPORTED_EMBEDDINGS = {
    en: {
        glove: [50],
        '10k-fasttext': [300],
    },
    ca: {
        '10k-fasttext': [300],
    },
    de: {
        '10k-fasttext': [300],
    },
    es: {
        '10k-fasttext': [300],
    },
    fr: {
        '10k-fasttext': [300],
    },
    hi: {
        '10k-fasttext': [300],
    },
    id: {
        '10k-fasttext': [300],
    },
    it: {
        '10k-fasttext': [300],
    },
    pt: {
        '10k-fasttext': [300],
    },
    ru: {
        '10k-fasttext': [300],
    },
    tr: {
        '10k-fasttext': [300],
    },
    zh: {
        '10k-fasttext': [300],
    },
};
exports.GLOBAL_EMBEDDINGS_PATH = path_1.join(os_1.homedir(), constants_1.BOTONIC_GLOBAL_DIRNAME, constants_1.BOTONIC_GLOBAL_EMBEDDINGS_DIRNAME);
exports.EMBEDDINGS_URL = 'https://s3-eu-west-1.amazonaws.com/word-embeddings.hubtype.com';
exports.DB_TABLE_NAME = 'embeddings';
exports.DB_COLUMN_NAME = 'token';
