// @botonic/react/lib/node-app.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NodeApp = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _server = require("react-dom/server");

var _reactBot = require("./react-bot");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var NodeApp = /*#__PURE__*/function () {
  function NodeApp(options) {
    var _this = this;

    (0, _classCallCheck2["default"])(this, NodeApp);
    this.bot = new _reactBot.ReactBot(_objectSpread({
      renderer: function renderer(args) {
        return _this.renderNode(args);
      }
    }, options));
  }

  (0, _createClass2["default"])(NodeApp, [{
    key: "renderNode",
    value: function () {
      var _renderNode = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(args) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.bot.renderReactActions(args);

              case 2:
                return _context.abrupt("return", _context.sent.map(function (action) {
                  return (0, _server.renderToStaticMarkup)(action);
                }).join('\n'));

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function renderNode(_x) {
        return _renderNode.apply(this, arguments);
      }

      return renderNode;
    }()
  }, {
    key: "input",
    value: function input(args) {
      return this.bot.input(args);
    }
  }, {
    key: "getConfig",
    value: function getConfig() {
      return Object.entries(this.bot.plugins).map(function (_ref) {
        var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
            _ = _ref2[0],
            plugin = _ref2[1];

        return {
          id: plugin.id,
          name: plugin.name,
          config: plugin.config
        };
      });
    }
  }]);
  return NodeApp;
}();

exports.NodeApp = NodeApp;
