// @botonic/react/lib/webview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebviewApp = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _core = require("@botonic/core");

var _axios = _interopRequireDefault(require("axios"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = require("react-dom");

var _reactRouterDom = require("react-router-dom");

var _contexts = require("./contexts");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var App = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(App, _React$Component);

  var _super = _createSuper(App);

  function App(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, App);
    _this = _super.call(this, props);
    var url = new URL(window.location.href);
    var params = Array.from(url.searchParams.entries()).filter(function (_ref) {
      var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
          key = _ref2[0],
          value = _ref2[1];

      return key != 'context';
    }).reduce(function (o, _ref3) {
      var _ref4 = (0, _slicedToArray2["default"])(_ref3, 2),
          key = _ref4[0],
          value = _ref4[1];

      o[key] = value;
      return o;
    }, {});
    var session = JSON.parse(url.searchParams.get('context') || {});
    _this.state = {
      session: session,
      params: params
    };
    return _this;
  }

  (0, _createClass2["default"])(App, [{
    key: "close",
    value: function () {
      var _close = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(options) {
        var payload, s, baseUrl, resp, provider, impId, phone_number;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                payload = options ? options.payload : null;
                if (options.path) payload = "__PATH_PAYLOAD__".concat(options.path);

                if (!payload) {
                  _context.next = 15;
                  break;
                }

                if (options.params) {
                  payload = "".concat(payload, "?").concat((0, _core.params2queryString)(options.params));
                }

                s = this.state.session;
                _context.prev = 5;
                baseUrl = s._hubtype_api || 'https://api.hubtype.com';
                _context.next = 9;
                return (0, _axios["default"])({
                  method: 'post',
                  url: "".concat(baseUrl, "/v1/bots/").concat(s.bot.id, "/send_postback/"),
                  // eslint-disable-next-line @typescript-eslint/naming-convention
                  data: {
                    payload: payload,
                    chat_id: s.user.id
                  }
                });

              case 9:
                resp = _context.sent;
                _context.next = 15;
                break;

              case 12:
                _context.prev = 12;
                _context.t0 = _context["catch"](5);
                console.log(_context.t0);

              case 15:
                provider = this.state.session.user.provider;
                impId = this.state.session.user.imp_id;

                if (provider === _core.PROVIDER.WHATSAPP) {
                  phone_number = this.state.session.user.unformatted_phone_number;
                  location.href = 'https://wa.me/' + phone_number;
                }

                if (provider === _core.PROVIDER.TELEGRAM) {
                  location.href = 'https://t.me/' + impId;
                }

                if (provider === _core.PROVIDER.APPLE) {
                  location.href = 'https://bcrw.apple.com/urn:biz:' + impId;
                }

                if (provider === _core.PROVIDER.TWITTER) {
                  location.href = 'https://twitter.com/messages/compose?recipient_id=' + impId;
                }

                if (provider === _core.PROVIDER.FACEBOOK) {
                  try {
                    window.MessengerExtensions.requestCloseBrowser(function () {
                      return undefined;
                    }, function (err) {
                      return console.log(err);
                    });
                  } catch (e) {}
                }

                if (!(provider === _core.PROVIDER.WEBCHAT)) {
                  _context.next = 30;
                  break;
                }

                _context.prev = 23;
                _context.next = 26;
                return parent.postMessage('botonicCloseWebview', '*');

              case 26:
                _context.next = 30;
                break;

              case 28:
                _context.prev = 28;
                _context.t1 = _context["catch"](23);

              case 30:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[5, 12], [23, 28]]);
      }));

      function close(_x) {
        return _close.apply(this, arguments);
      }

      return close;
    }()
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var requestContext = {
        getString: function getString(stringId) {
          return (0, _core.getString)(_this2.props.locales, _this2.state.session.__locale, stringId);
        },
        session: this.state.session || {},
        params: this.state.params || {},
        closeWebview: this.close.bind(this)
      };
      return /*#__PURE__*/_react["default"].createElement(_contexts.RequestContext.Provider, {
        value: requestContext
      }, this.props.webviews.map(function (Webview, i) {
        return /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Route, {
          key: i,
          path: "/".concat(Webview.name),
          component: Webview
        });
      }));
    }
  }]);
  return App;
}(_react["default"].Component);

var WebviewApp = /*#__PURE__*/function () {
  function WebviewApp(_ref5) {
    var webviews = _ref5.webviews,
        locales = _ref5.locales;
    (0, _classCallCheck2["default"])(this, WebviewApp);
    this.webviews = webviews;
    this.locales = locales;
  }

  (0, _createClass2["default"])(WebviewApp, [{
    key: "render",
    value: function render(dest) {
      (0, _reactDom.render)( /*#__PURE__*/_react["default"].createElement(_reactRouterDom.BrowserRouter, null, /*#__PURE__*/_react["default"].createElement(App, {
        webviews: this.webviews,
        locales: this.locales
      })), dest);
    }
  }]);
  return WebviewApp;
}();

exports.WebviewApp = WebviewApp;
