// @botonic/core/lib/esm/models/legacy-types.js
// TODO: This file contains all the legacy types we had in index.ts. After some refactors, we should be able to get rid of many of them.
export const PROVIDER = Object.freeze({
    APPLE: 'apple',
    DEV: 'dev',
    FACEBOOK: 'facebook',
    GENERIC: 'generic',
    INSTAGRAM: 'instagram',
    INTERCOM: 'intercom',
    SMOOCH: 'smooch',
    TELEGRAM: 'telegram',
    TWITTER: 'twitter',
    WEBCHAT: 'webchat',
    WECHAT: 'wechat',
    WHATSAPP: 'whatsapp',
});
export const INPUT = Object.freeze({
    TEXT: 'text',
    POSTBACK: 'postback',
    AUDIO: 'audio',
    IMAGE: 'image',
    VIDEO: 'video',
    DOCUMENT: 'document',
    LOCATION: 'location',
    CONTACT: 'contact',
    BUTTON_MESSAGE: 'buttonmessage',
    CAROUSEL: 'carousel',
    CUSTOM: 'custom',
    WEBCHAT_SETTINGS: 'webchatsettings',
    WHATSAPP_TEMPLATE: 'whatsapptemplate',
    RAW: 'raw',
    CHAT_EVENT: 'chatevent',
});
export const CASE_STATUS = Object.freeze({
    WAITING: 'status_waiting',
    ATTENDING: 'status_attending',
    IDLE: 'status_idle',
    RESOLVED: 'status_resolved',
});
export const CASE_RESOLUTION = Object.freeze({
    OK: 'result_ok',
    NOK: 'result_nok',
    NOT_SOLVED: 'result_not_solved',
    BANNED: 'result_banned',
});
export const Providers = Object.freeze({
    Messaging: {
        APPLE: 'apple',
        FACEBOOK: 'facebook',
        GENERIC: 'generic',
        IMBEE: 'imbee',
        INSTAGRAM: 'instagram',
        INTERCOM: 'intercom',
        SMOOCH_WEB: 'smooch_web',
        SMOOCH: 'smooch',
        TELEGRAM: 'telegram',
        TWITTER: 'twitter',
        WEBCHAT: 'webchat',
        WECHAT: 'wechat',
        WHATSAPP: 'whatsapp',
    },
});
