// @botonic/nlp/lib/dataset/input-data-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputDataReader = void 0;
var fs_1 = require("fs");
var js_yaml_1 = require("js-yaml");
var path_1 = require("path");
var InputDataReader = /** @class */ (function () {
    function InputDataReader(path) {
        this.path = path;
        this.ALLOWED_EXTENSIONS = ['.yaml', '.yml'];
        this.filePaths = this.getInputFilePaths(path);
    }
    InputDataReader.prototype.read = function () {
        return this.filePaths.map(function (path) { return js_yaml_1.load(fs_1.readFileSync(path)); });
    };
    InputDataReader.prototype.getInputFilePaths = function (path) {
        var _this = this;
        var stat = fs_1.lstatSync(path);
        if (!stat.isDirectory()) {
            throw new Error("path '" + path + "' must be a directory.");
        }
        return fs_1.readdirSync(path)
            .filter(function (fileName) { return _this.ALLOWED_EXTENSIONS.includes(path_1.extname(fileName)); })
            .map(function (fileName) { return path_1.join(path, fileName); });
    };
    return InputDataReader;
}());
exports.InputDataReader = InputDataReader;
