// @botonic/nlp/lib/encode/indexed-items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexedItems = void 0;
var array_utils_1 = require("../utils/array-utils");
var IndexedItems = /** @class */ (function () {
    function IndexedItems(items) {
        this.items = array_utils_1.unique(items);
    }
    IndexedItems.prototype.includes = function (item) {
        return this.items.includes(item);
    };
    IndexedItems.prototype.getIndex = function (item) {
        var idx = this.items.indexOf(item);
        if (idx === -1) {
            throw new Error("Item '" + item + "' not found.");
        }
        return idx;
    };
    IndexedItems.prototype.getItem = function (idx) {
        if (!this.isValidIndex(idx)) {
            throw new Error("Invalid index '" + idx + "'.");
        }
        return this.items[idx];
    };
    IndexedItems.prototype.maskUnknownItems = function (sequence, unknownItem) {
        var _this = this;
        return sequence.map(function (item) {
            return _this.items.includes(item) ? item : unknownItem;
        });
    };
    IndexedItems.prototype.isValidIndex = function (idx) {
        return 0 <= idx && idx < this.length;
    };
    Object.defineProperty(IndexedItems.prototype, "length", {
        get: function () {
            return this.items.length;
        },
        enumerable: false,
        configurable: true
    });
    return IndexedItems;
}());
exports.IndexedItems = IndexedItems;
