// @botonic/nlp/lib/encode/one-hot-encoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneHotEncoder = void 0;
var OneHotEncoder = /** @class */ (function () {
    function OneHotEncoder(items) {
        this.items = items;
    }
    OneHotEncoder.prototype.encode = function (sequence) {
        var _this = this;
        return sequence.map(function (item) { return _this.encodeItem(item); });
    };
    OneHotEncoder.prototype.encodeItem = function (item) {
        var idx = this.items.getIndex(item);
        return this.indexToOneHotVector(idx);
    };
    OneHotEncoder.prototype.indexToOneHotVector = function (idx) {
        var vector = Array(this.items.length).fill(0);
        vector[idx] = 1;
        return vector;
    };
    OneHotEncoder.prototype.decode = function (sequence) {
        var _this = this;
        return sequence.map(function (vector) { return _this.decodeOneHotVector(vector); });
    };
    OneHotEncoder.prototype.decodeOneHotVector = function (vector) {
        var idx = this.oneHotVectorToIndex(vector);
        return this.items.getItem(idx);
    };
    OneHotEncoder.prototype.oneHotVectorToIndex = function (vector) {
        if (vector.length !== this.items.length) {
            throw new Error("Invalid categorical length '" + vector.length + "'.");
        }
        return vector.indexOf(Math.max.apply(Math, vector));
    };
    return OneHotEncoder;
}());
exports.OneHotEncoder = OneHotEncoder;
