// @botonic/core/lib/cjs/models/events/message/message-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageEventFrom = exports.MessageEventAck = exports.MessageEventTypes = void 0;
var MessageEventTypes;
(function (MessageEventTypes) {
    MessageEventTypes["AUDIO"] = "audio";
    MessageEventTypes["CAROUSEL"] = "carousel";
    MessageEventTypes["CUSTOM"] = "custom";
    MessageEventTypes["DOCUMENT"] = "document";
    MessageEventTypes["IMAGE"] = "image";
    MessageEventTypes["LOCATION"] = "location";
    MessageEventTypes["TEXT"] = "text";
    MessageEventTypes["POSTBACK"] = "postback";
    MessageEventTypes["VIDEO"] = "video";
    /**
     * TODO: contact, buttonmessage?, webchatsettings, whatsapp template
     */
})(MessageEventTypes = exports.MessageEventTypes || (exports.MessageEventTypes = {}));
var MessageEventAck;
(function (MessageEventAck) {
    MessageEventAck["DRAFT"] = "draft";
    MessageEventAck["READ"] = "read";
    MessageEventAck["RECEIVED"] = "received";
    MessageEventAck["SENT"] = "sent";
})(MessageEventAck = exports.MessageEventAck || (exports.MessageEventAck = {}));
var MessageEventFrom;
(function (MessageEventFrom) {
    MessageEventFrom["AGENT"] = "agent";
    MessageEventFrom["BOT"] = "bot";
    MessageEventFrom["USER"] = "user";
})(MessageEventFrom = exports.MessageEventFrom || (exports.MessageEventFrom = {}));
