// @botonic/react/lib/components/audio.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Audio = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _core = require("@botonic/core");

var _react = _interopRequireDefault(require("react"));

var _constants = require("../constants");

var _message = require("./message");

var serialize = function serialize(audioProps) {
  return {
    audio: audioProps.src
  };
};

var Audio = function Audio(props) {
  var content = props.children;
  if ((0, _core.isBrowser)()) content = /*#__PURE__*/_react["default"].createElement("audio", {
    style: {
      maxWidth: '100%'
    },
    id: "myAudio",
    controls: true
  }, /*#__PURE__*/_react["default"].createElement("source", {
    src: props.src,
    type: "audio/mpeg"
  }), "Your browser does not support this audio format.");
  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    role: _constants.ROLES.AUDIO_MESSAGE,
    json: serialize(props)
  }, props, {
    type: _core.INPUT.AUDIO
  }), content);
};

exports.Audio = Audio;
Audio.serialize = serialize;
