// @botonic/react/lib/components/custom-message.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.customMessage = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _core = require("@botonic/core");

var _lodash = _interopRequireDefault(require("lodash.merge"));

var _react = _interopRequireDefault(require("react"));

var _constants = require("../constants");

var _errorBoundary = require("../util/error-boundary");

var _logs = require("../util/logs");

var _react2 = require("../util/react");

var _message = require("./message");

var _reply = require("./reply");

var _excluded = ["id", "children"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

/**
 *
 * @param name as it appears at ThemeProps' message.customTypes key
 * @param CustomMessageComponent
 * @param defaultProps Props for the wrapper Message
 * @param ErrorBoundary to recover in case it fails
 */
var customMessage = function customMessage(_ref) {
  var name = _ref.name,
      CustomMessageComponent = _ref.component,
      _ref$defaultProps = _ref.defaultProps,
      defaultProps = _ref$defaultProps === void 0 ? {} : _ref$defaultProps,
      _ref$errorBoundary = _ref.errorBoundary,
      ErrorBoundary = _ref$errorBoundary === void 0 ? (0, _errorBoundary.createErrorBoundary)() : _ref$errorBoundary;

  var CustomMessage = function CustomMessage(props) {
    (0, _logs.warnDeprecatedProps)(defaultProps, 'customMessage:');
    if (defaultProps.from === _constants.SENDERS.user) defaultProps.ack = 1;
    return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({}, (0, _lodash["default"])((0, _react2.mapObjectNonBooleanValues)(defaultProps), props), {
      type: _core.INPUT.CUSTOM
    }));
  };

  var splitChildren = function splitChildren(props) {
    var children = props.children;

    var isReply = function isReply(e) {
      return e.type === _reply.Reply;
    };

    try {
      if (!Array.isArray(children) && !isReply(children)) {
        return {
          replies: null,
          childrenWithoutReplies: children
        };
      }

      var childrenArray = _react["default"].Children.toArray(children);

      var replies = childrenArray.filter(isReply);
      var childrenWithoutReplies = childrenArray.filter(function (e) {
        return !isReply(e);
      });
      return {
        replies: replies,
        childrenWithoutReplies: childrenWithoutReplies
      };
    } catch (e) {
      return {
        replies: null,
        childrenWithoutReplies: children
      };
    }
  };

  var WrappedComponent = function WrappedComponent(props) {
    var id = props.id,
        children = props.children,
        customMessageProps = (0, _objectWithoutProperties2["default"])(props, _excluded);

    var _splitChildren = splitChildren(props),
        replies = _splitChildren.replies,
        childrenWithoutReplies = _splitChildren.childrenWithoutReplies;

    return /*#__PURE__*/_react["default"].createElement(CustomMessage, {
      id: id,
      json: _objectSpread(_objectSpread({}, customMessageProps), {}, {
        id: id,
        children: childrenWithoutReplies,
        customTypeName: name
      })
    }, /*#__PURE__*/_react["default"].createElement(ErrorBoundary, (0, _extends2["default"])({
      key: 'errorBoundary'
    }, customMessageProps), /*#__PURE__*/_react["default"].createElement(CustomMessageComponent, customMessageProps, childrenWithoutReplies)), replies);
  };

  WrappedComponent.customTypeName = name; // eslint-disable-next-line react/display-name

  WrappedComponent.deserialize = function (msg) {
    return /*#__PURE__*/_react["default"].createElement(WrappedComponent, (0, _extends2["default"])({
      id: msg.id,
      key: msg.key,
      json: msg.data
    }, msg.data));
  };

  return WrappedComponent;
};

exports.customMessage = customMessage;
