// @botonic/react/lib/components/image.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Image = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@botonic/core");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _contexts = require("../contexts");

var _message = require("./message");

var _templateObject;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var StyledImage = _styledComponents["default"].img(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  border-radius: 8px;\n  max-width: 150px;\n  max-height: 150px;\n  margin: -3px -6px;\n  cursor: ", ";\n"])), function (_ref) {
  var hasPreviewer = _ref.hasPreviewer;
  return hasPreviewer ? 'pointer' : 'auto';
});

var serialize = function serialize(imageProps) {
  return {
    image: imageProps.src
  };
};

var Image = function Image(props) {
  var content = props.children;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      isPreviewerOpened = _useState2[0],
      setIsPreviewerOpened = _useState2[1];

  var openPreviewer = function openPreviewer() {
    return setIsPreviewerOpened(true);
  };

  var closePreviewer = function closePreviewer() {
    return setIsPreviewerOpened(false);
  };

  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      getThemeProperty = _useContext.getThemeProperty;

  var ImagePreviewer = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.imagePreviewer, null);

  if ((0, _core.isBrowser)()) {
    content = /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(StyledImage, {
      src: props.src,
      onClick: openPreviewer,
      hasPreviewer: Boolean(ImagePreviewer)
    }), ImagePreviewer && /*#__PURE__*/_react["default"].createElement(ImagePreviewer, {
      src: props.src,
      isPreviewerOpened: isPreviewerOpened,
      openPreviewer: openPreviewer,
      closePreviewer: closePreviewer
    }));
  }

  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    role: _constants.ROLES.IMAGE_MESSAGE,
    json: serialize(props)
  }, props, {
    type: _core.INPUT.IMAGE
  }), content);
};

exports.Image = Image;
Image.serialize = serialize;
