// @botonic/react/lib/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Audio: true,
  Button: true,
  Carousel: true,
  customMessage: true,
  Document: true,
  Element: true,
  Image: true,
  Location: true,
  MessageTemplate: true,
  Pic: true,
  Raw: true,
  Reply: true,
  ShareButton: true,
  Subtitle: true,
  Text: true,
  Title: true,
  Video: true,
  WebchatSettings: true,
  WhatsappTemplate: true
};
Object.defineProperty(exports, "Audio", {
  enumerable: true,
  get: function get() {
    return _audio.Audio;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _button.Button;
  }
});
Object.defineProperty(exports, "Carousel", {
  enumerable: true,
  get: function get() {
    return _carousel.Carousel;
  }
});
Object.defineProperty(exports, "Document", {
  enumerable: true,
  get: function get() {
    return _document.Document;
  }
});
Object.defineProperty(exports, "Element", {
  enumerable: true,
  get: function get() {
    return _element.Element;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function get() {
    return _image.Image;
  }
});
Object.defineProperty(exports, "Location", {
  enumerable: true,
  get: function get() {
    return _location.Location;
  }
});
Object.defineProperty(exports, "MessageTemplate", {
  enumerable: true,
  get: function get() {
    return _messageTemplate.MessageTemplate;
  }
});
Object.defineProperty(exports, "Pic", {
  enumerable: true,
  get: function get() {
    return _pic.Pic;
  }
});
Object.defineProperty(exports, "Raw", {
  enumerable: true,
  get: function get() {
    return _raw.Raw;
  }
});
Object.defineProperty(exports, "Reply", {
  enumerable: true,
  get: function get() {
    return _reply.Reply;
  }
});
Object.defineProperty(exports, "ShareButton", {
  enumerable: true,
  get: function get() {
    return _shareButton.ShareButton;
  }
});
Object.defineProperty(exports, "Subtitle", {
  enumerable: true,
  get: function get() {
    return _subtitle.Subtitle;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function get() {
    return _text.Text;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function get() {
    return _title.Title;
  }
});
Object.defineProperty(exports, "Video", {
  enumerable: true,
  get: function get() {
    return _video.Video;
  }
});
Object.defineProperty(exports, "WebchatSettings", {
  enumerable: true,
  get: function get() {
    return _webchatSettings.WebchatSettings;
  }
});
Object.defineProperty(exports, "WhatsappTemplate", {
  enumerable: true,
  get: function get() {
    return _whatsappTemplate.WhatsappTemplate;
  }
});
Object.defineProperty(exports, "customMessage", {
  enumerable: true,
  get: function get() {
    return _customMessage.customMessage;
  }
});

var _audio = require("./audio");

var _button = require("./button");

var _carousel = require("./carousel");

var _customMessage = require("./custom-message");

var _document = require("./document");

var _element = require("./element");

var _image = require("./image");

var _location = require("./location");

var _messageTemplate = require("./message-template");

var _multichannel = require("./multichannel");

Object.keys(_multichannel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _multichannel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _multichannel[key];
    }
  });
});

var _pic = require("./pic");

var _raw = require("./raw");

var _reply = require("./reply");

var _shareButton = require("./share-button");

var _subtitle = require("./subtitle");

var _text = require("./text");

var _title = require("./title");

var _video = require("./video");

var _webchatSettings = require("./webchat-settings");

var _whatsappTemplate = require("./whatsapp-template");
