// @botonic/react/lib/components/location.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Location = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@botonic/core");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _react2 = require("../util/react");

var _message = require("./message");

var _templateObject;

var Link = _styledComponents["default"].a(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  text-decoration: none;\n  font-weight: bold;\n  target: blank;\n"])));

var serialize = function serialize(locationProps) {
  return {
    location: {
      lat: locationProps.lat,
      "long": locationProps["long"]
    }
  };
};

var Location = function Location(props) {
  var lat = parseFloat(props.lat);

  var _long = parseFloat(props["long"]);

  var renderBrowser = function renderBrowser() {
    var locationUrl = "https://www.google.com/maps/search/?api=1&query=".concat(lat, ",").concat(_long);
    return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
      json: serialize(props)
    }, props, {
      type: _core.INPUT.LOCATION
    }), /*#__PURE__*/_react["default"].createElement(Link, {
      href: locationUrl,
      target: "_blank",
      rel: "noopener noreferrer"
    }, props.text || 'Open Location'));
  };

  var renderNode = function renderNode() {
    return /*#__PURE__*/_react["default"].createElement("message", {
      type: _core.INPUT.LOCATION
    }, /*#__PURE__*/_react["default"].createElement("lat", null, lat), /*#__PURE__*/_react["default"].createElement("long", null, _long));
  };

  return (0, _react2.renderComponent)({
    renderBrowser: renderBrowser,
    renderNode: renderNode
  });
};

exports.Location = Location;
Location.serialize = serialize;
