// @botonic/react/lib/components/raw.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Raw = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@botonic/core");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _message = require("./message");

var _templateObject, _templateObject2;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var RawDataButton = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  align-self: flex-start;\n  background-color: ", ";\n  padding: 4px;\n  margin: 4px 0px;\n  cursor: pointer;\n  color: ", ";\n  font-size: 10px;\n  border-radius: 2px;\n"])), _constants.COLORS.CURIOUS_BLUE, _constants.COLORS.SOLID_WHITE_ALPHA_0_8);

var Popover = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  max-width: 100%;\n  max-height: 500px;\n  overflow: auto;\n  background-color: ", ";\n  padding: 10px;\n  pre {\n    margin: 0px;\n  }\n"])), _constants.COLORS.LIGHT_GRAY);

var serialize = function serialize(rawProps) {
  return {
    data: rawProps.data,
    alt: rawProps.alt
  };
};

var Raw = function Raw(props) {
  var content = props.children;
  var data = JSON.stringify(props.data, null, ' ');

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      isOpen = _useState2[0],
      setIsOpen = _useState2[1];

  if ((0, _core.isBrowser)()) content = /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("div", null, props.alt), /*#__PURE__*/_react["default"].createElement(RawDataButton, {
    onClick: function onClick() {
      return setIsOpen(!isOpen);
    }
  }, "RAW DATA ", /*#__PURE__*/_react["default"].createElement("small", null, isOpen ? '▲' : '▼')), isOpen && /*#__PURE__*/_react["default"].createElement(Popover, null, /*#__PURE__*/_react["default"].createElement("pre", null, data)));
  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    role: _constants.ROLES.RAW_MESSAGE,
    json: serialize(props)
  }, props, {
    data: data,
    type: _core.INPUT.RAW
  }), content);
};

exports.Raw = Raw;
Raw.serialize = serialize;
