// @botonic/react/lib/components/timestamps.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resolveMessageTimestamps = exports.MessageTimestamp = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _templateObject, _templateObject2;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var resolveMessageTimestamps = function resolveMessageTimestamps(getThemeFn, messageTimestampEnabled) {
  var timestampsFormat = getThemeFn('message.timestamps.format');
  var timestampStyle = getThemeFn('message.timestamps.style');
  var timestampsEnabled = Boolean(messageTimestampEnabled !== undefined ? messageTimestampEnabled : getThemeFn('message.timestamps.enable', Boolean(timestampStyle) || Boolean(timestampsFormat) || false));
  var defaultTimestampFormat = {
    month: 'short',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric'
  };
  var getFormattedTimestamp = timestampsFormat && timestampsFormat() || new Date().toLocaleString(undefined, defaultTimestampFormat);
  return {
    timestampsEnabled: timestampsEnabled,
    getFormattedTimestamp: getFormattedTimestamp,
    timestampStyle: timestampStyle
  };
};

exports.resolveMessageTimestamps = resolveMessageTimestamps;

var TimestampContainer = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  position: relative;\n  align-items: flex-start;\n"])));

var TimestampText = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  @import url('https://fonts.googleapis.com/css?family=Lato');\n  font-family: Lato;\n  font-size: 10px;\n  color: ", ";\n  width: 100%;\n  text-align: ", ";\n  padding: ", ";\n  margin-bottom: 5px;\n"])), _constants.COLORS.SOLID_BLACK, function (props) {
  return props.isfromuser ? 'right' : 'left';
}, function (props) {
  return props.isfromuser ? '0px 15px' : '0px 50px';
});

var MessageTimestamp = function MessageTimestamp(_ref) {
  var timestamp = _ref.timestamp,
      style = _ref.style,
      isfromuser = _ref.isfromuser;
  var classSufix = isfromuser ? 'user' : 'bot';
  return /*#__PURE__*/_react["default"].createElement(TimestampContainer, {
    className: "botonic-timestamp-".concat(classSufix)
  }, /*#__PURE__*/_react["default"].createElement(TimestampText, {
    isfromuser: isfromuser,
    style: _objectSpread({}, style)
  }, timestamp));
};

exports.MessageTimestamp = MessageTimestamp;
