// @botonic/react/lib/components/video.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Video = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@botonic/core");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _message = require("./message");

var _templateObject;

var StyledVideo = _styledComponents["default"].video(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: ", ";\n  border-radius: 8px;\n  max-height: 180px;\n  max-width: 300px;\n  margin: 10px;\n"])), _constants.COLORS.SOLID_BLACK_ALPHA_0_5);

var serialize = function serialize(videoProps) {
  return {
    video: videoProps.src
  };
};

var Video = function Video(props) {
  var content = props.children;
  if ((0, _core.isBrowser)()) content = /*#__PURE__*/_react["default"].createElement(StyledVideo, {
    controls: true
  }, /*#__PURE__*/_react["default"].createElement("source", {
    src: props.src
  }));
  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    role: _constants.ROLES.VIDEO_MESSAGE,
    json: serialize(props)
  }, props, {
    type: _core.INPUT.VIDEO
  }), content);
};

exports.Video = Video;
Video.serialize = serialize;
