// @botonic/react/lib/components/whatsapp-template.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WhatsappTemplate = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _core = require("@botonic/core");

var _react = _interopRequireDefault(require("react"));

var _react2 = require("../util/react");

var _message = require("./message");

var serialize = function serialize(whatsappTemplateProps) {
  return {
    text: whatsappTemplateProps
  };
};

var WhatsappTemplate = function WhatsappTemplate(props) {
  var renderBrowser = function renderBrowser() {
    // Return a dummy message for browser
    var message = "Template ".concat(props.name, " with namespace ").concat(props.namespace, " would be sent to the user.");
    return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
      json: serialize(message)
    }, props, {
      type: _core.INPUT.TEXT
    }), message);
  };

  var renderNode = function renderNode() {
    return /*#__PURE__*/_react["default"].createElement("message", (0, _extends2["default"])({}, props, {
      header: props.header && JSON.stringify(props.header),
      body: props.body && JSON.stringify(props.body),
      footer: props.footer && JSON.stringify(props.footer),
      type: _core.INPUT.WHATSAPP_TEMPLATE
    }));
  };

  return (0, _react2.renderComponent)({
    renderBrowser: renderBrowser,
    renderNode: renderNode
  });
};

exports.WhatsappTemplate = WhatsappTemplate;
