// @botonic/core/lib/esm/data-provider/dynamodb-data-provider.js
import { EventTypes } from '../models';
import { getConnectionEventEntity, getMessageEventEntities, getUserEntity, getUserEventsTable, GLOBAL_SECONDARY_INDEX_NAME, SORT_KEY_NAME, USER_PREFIX, } from './dynamodb-utils';
export class DynamoDBDataProvider {
    constructor(url) {
        try {
            ;
            [this.tableName, this.region] = url.split('://')[1].split('.');
            this.userEventsTable = getUserEventsTable(this.tableName, this.region);
            this.userEntity = getUserEntity(this.userEventsTable);
            this.connectionEventEntity = getConnectionEventEntity(this.userEventsTable);
            this.messageEventEntities = getMessageEventEntities(this.userEventsTable);
        }
        catch (e) {
            console.log({ e });
        }
    }
    // @ts-ignore
    async getUsers(limit = 10, offset = 0) {
        // TODO: finish to implement with offset
        const result = await this.userEventsTable.scan({
            filters: { attr: SORT_KEY_NAME, beginsWith: USER_PREFIX },
            limit,
        });
        return result.Items;
    }
    async getUser(id) {
        const userById = {
            id: id,
            SK: id,
        };
        const result = await this.userEntity.get(userById);
        if (Object.keys(result).length === 0)
            return undefined;
        return result.Item;
    }
    async getUserByWebsocketId(websocketId) {
        const result = await this.userEventsTable.query(websocketId, {
            index: GLOBAL_SECONDARY_INDEX_NAME,
        });
        if (result.Count === 0)
            return undefined;
        return result.Items[0];
    }
    // @ts-ignore
    async getUserByField(field, value) { } //TODO: Implement
    async saveUser(user) {
        const putUser = Object.assign(Object.assign({}, user), { id: user.id, userId: user.id });
        await this.userEntity.put(putUser);
        return user;
    }
    async updateUser(user) {
        const updateUser = Object.assign(Object.assign({}, user), { id: user.id, userId: user.id });
        const res = await this.userEntity.update(updateUser);
        return user;
    }
    // @ts-ignore
    async deleteUser(id) { } // TODO: Implement
    // @ts-ignore
    async getEvents(limit = 10, offset = 0) { } // TODO: Implement
    // @ts-ignore
    async getEvent(id) { } // TODO: Implement
    async saveEvent(event) {
        if (event.eventType === EventTypes.CONNECTION) {
            await this.connectionEventEntity.put(event);
        }
        if (event.eventType === EventTypes.MESSAGE) {
            await this.messageEventEntities[event.type].put(event);
        }
        // TODO: Implementation of other event types
        return event;
    }
    // @ts-ignore
    async updateEvent(event) { } // TODO: Implement
    // @ts-ignore
    async deleteEvent(id) { } // TODO: Implement
}
